table data iter gains new members next_paged(), get_schema() and get_rows_fetched()

krlmlr opened the pull request:
Allows fetching all rows from a query, extracted DEFAULT_PAGE_SIZE as private variable.

craigcitro commented on R/tabledata.r:
```diff
@@ -131,7 +133,16 @@ list_tabledata_iter <- function(project, dataset, table, table_info = NULL) {
   }
 
-  list(next_ = next_, is_complete = is_complete)
+  all <- function(page_size = DEFAULT_PAGE_SIZE) {
```
it might be worth taking this as an argument to `list_tabledata_iter`, even if just for testing purposes.

krlmlr replied:
Good point. I'll convert this to a functions, which then can be mocked by tests.

craigcitro replied:
Ah, I actually meant something slightly different -- make this an explicit argument to `list_tabledata_iter`, and then pass an explicit value in tests. (it's also sometimes useful for end-users, though more rare.)

craigcitro commented:
One small nit, and the standard question -- can we add a test? :grin:

krlmlr commented:
#14.

krlmlr commented:
default_page_size() is now a function. Also added two more members.

krlmlr commented:
re https://github.com/rstats-db/bigrquery/pull/87#discussion_r52267196: I think that it's a bit too much to support an argument to the outer function that specifies the *default* value (which can be overridden in all() anyway if necessary). For testing, default_page_size() can be mocked.

hadley commented on R/tabledata.r:
```diff
@@ -52,7 +54,8 @@ list_tabledata <- function(project, dataset, table, page_size = 1e4,
 list_tabledata_callback <- function(project, dataset, table, callback,
                                     table_info = NULL,
-                                    page_size = 1e4, max_pages = 10,
+                                    page_size = default_page_size(),
```
I don't like using non-exported functions for defaults because it's hard to find out what the default value is

krlmlr replied:
I'll add an option instead.

hadley commented on R/tabledata.r:
```diff
@@ -131,7 +134,25 @@ list_tabledata_iter <- function(project, dataset, table, table_info = NULL) {
-  list(next_ = next_, is_complete = is_complete)
+  all <- function(page_size = default_page_size()) {
+    ret <- list()
+    while (!is_complete()) {
```
This feels a bit icky, but I think doing it right will be a lot of work and is unlikely to have practical performance implication. Instead add a comment that says: "Fetching large data from bigrquery will be slow for other reasons..."

krlmlr commented:
@craigcitro: The all() function is needed for DBI, and will be tested there implicitly. We probably want to set the default page size to something smallish there.

krlmlr commented:
Now using an option to define default page size, plus other enhancements. @craigcitro: PTAL.

hadley commented:
LGTM

craigcitro commented:
Just to check, is the diff here picking up anything extra? (In particular, the extra note at the top of NEWS looks like it should be on another PR?)

krlmlr commented:
The NEWS item was missing from another PR, I added it here for simplicity. Besides that, all changes relate to the list_tabledata_iter() function; I've also switched from integer to numeric to account for results with more than 2G rows.
